consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/MessageValidator.java

```diff
@@ -17,20 +17,63 @@
 import org.hyperledger.besu.consensus.qbft.messagewrappers.Commit;
 import org.hyperledger.besu.consensus.qbft.messagewrappers.Prepare;
 import org.hyperledger.besu.consensus.qbft.messagewrappers.Proposal;
+import org.hyperledger.besu.ethereum.core.Hash;
+
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public class MessageValidator {
 
-  public MessageValidator() {}
+  private static final Logger LOG = LogManager.getLogger();
+
+  @FunctionalInterface
+  public interface PrepareValidatorFactory {
+    PrepareValidator create(final Hash expectedDigest);
+  }
+
+  @FunctionalInterface
+  public interface CommitValidatorFactory {
+    CommitValidator create(final Hash expectedDigest);
+  }
+
+  private final PrepareValidatorFactory prepareValidatorFactory;
+  private final CommitValidatorFactory commitValidatorFactory;
+  private final ProposalValidator proposalValidator;
+
+  private Optional<PrepareValidator> prepareValidator = Optional.empty();
+  private Optional<CommitValidator> commitValidator = Optional.empty();
+
+  public MessageValidator(
+      final PrepareValidatorFactory prepareValidatorFactory,
+      final CommitValidatorFactory commitValidatorFactory,
+      final ProposalValidator proposalValidator) {
+    this.prepareValidatorFactory = prepareValidatorFactory;
+    this.commitValidatorFactory = commitValidatorFactory;
+    this.proposalValidator = proposalValidator;
+  }
 
   public boolean validateProposal(final Proposal msg) {
-    return true;
+    if (prepareValidator.isPresent()) {
+      LOG.info("Received subsequent Proposal for current round, discarding.");
+      return false;
+    }
+
+    final boolean result = proposalValidator.validate(msg);
+    if (result) {
+      prepareValidator = Optional.of(prepareValidatorFactory.create(msg.getBlock().getHash()));
```

Comment:
Don't see why these are condition on the validation passing. They are dependent on having a block.